lib/Dialect/TritonGPU/Transforms/WarpSpecialization/PartitionLoops.cpp

```diff
@@ -16,103 +19,337 @@ using namespace mlir;
+    for (auto src : reduceOp.getSrcs()) {
+      srcs.push_back(b.mapping.lookupOrDefault(src));
+    }
+    auto axis = reduceOp.getAxis();
+    auto newReduceOp =
+        b.builder.create<triton::ReduceOp>(reduceOp.getLoc(), srcs, axis);
+    newReduceOp->setAttrs(reduceOp->getAttrs());
+    newReduceOps.push_back(newReduceOp);
+
+    for (auto [oldResult, newResult] :
+         llvm::zip(reduceOp.getResults(), newReduceOp.getResults())) {
+      b.mapping.map(oldResult, newResult);
+    }
+
+    auto &region = newReduceOp.getRegion();
+    Block *block = &region.emplaceBlock();
+    for (auto args : reduceOp.getRegion().getBlocks().front().getArguments()) {
+      auto newArg = block->addArgument(args.getType(), args.getLoc());
+      b.mapping.map(args, newArg);
+    }
+
+    b.builder.setInsertionPointToStart(block);
+  }
+
+  cloneOpsInBlock(reduceOp.getBody(), builders, schedule);
+
+  for (auto [idx, newReduceOp] : llvm::zip(partitionIndices, newReduceOps)) {
+    builders[idx].builder.setInsertionPointAfter(newReduceOp);
+  }
+}
+
+void cloneOpsInBlock(Block *block, SmallVector<WgBuilder> &builders,
+                     const WarpSchedule &schedule) {
+  for (auto &op_ : *block) {
+    auto op = &op_;
+    auto partition = getPartition(op, schedule);
+
+    if (auto forOp = dyn_cast<scf::ForOp>(op)) {
+      cloneForOp(forOp, builders, schedule);
+    } else if (auto ifOp = dyn_cast<scf::IfOp>(op)) {
+      cloneIfOp(ifOp, builders, schedule);
+    } else if (auto reduceOp = dyn_cast<triton::ReduceOp>(op)) {
+      cloneReduceOp(reduceOp, builders, schedule);
+    } else if (auto yieldOp = dyn_cast<scf::YieldOp>(op)) {
+      if (yieldOp.getOperands().empty()) {
+        continue;
+      }
+
+      auto doClone = [&](WgBuilder &builder) {
+        SmallVector<size_t> newOperandIndices;
+        if (auto forOp = dyn_cast<scf::ForOp>(yieldOp->getParentOp())) {
+          newOperandIndices =
+              getLoopVarIndicesToKeep(
+                  forOp, schedule.getPartition(builder.partitionId), schedule)
+                  .first;
+        } else {
+          for (size_t i = 0; i < yieldOp.getOperands().size(); ++i) {
+            newOperandIndices.push_back(i);
+          }
+        }
+
+        SmallVector<Value> newYieldOperands;
+        for (size_t i : newOperandIndices) {
+          newYieldOperands.push_back(
+              builder.mapping.lookupOrDefault(yieldOp.getOperand(i)));
+        }
+
+        if (!newYieldOperands.empty()) {
+          builder.builder.create<scf::YieldOp>(op->getLoc(), newYieldOperands);
+        }
+      };
+
+      if (!partition || partition == schedule.getRootPartition()) {
+        for (auto &builder : builders) {
+          doClone(builder);
+        }
+      } else {
+        auto &builder = builders[partition->getIndex()];
+        doClone(builder);
+      }
+    } else {
+      // all remaining ops are expected to be regionless
+      assert(op->getNumRegions() == 0);
```

Comment:
can you replace this with a `llvm::report_fatal_error`? It will help future compiler devs figure out what they need to add